engine/tests/phpunit/Elgg/EntityIconServiceTest.php

```diff
@@ -0,0 +1,455 @@
+
+		foreach ($sizes as $size => $opts) {
+			$this->assertFalse($service->hasIcon($this->entity, $size));
+		}
+	}
+
+	/**
+	 * @group IconService
+	 * @expectedException \InvalidParameterException
+	 */
+	public function testThrowsExceptionIfLocalFileIsNotReadable() {
+		$service = $this->createService();
+		$local_file = $this->config->get('dataroot') . '_______empty';
+		$service->saveIconFromLocalFile($this->entity, $local_file);
+	}
+
+	/**
+	 * @group IconService
+	 */
+	public function testCanSaveIconFromLocalFile() {
+
+		$service = $this->createService();
+
+		$sizes = $service->getSizes('object', 'foo');
+		$this->assertNotEmpty($sizes);
+
+		$local_file = $this->config->get('dataroot') . '1/1/white.png';
+		$service->saveIconFromLocalFile($this->entity, $local_file);
+
+		foreach ($sizes as $size => $opts) {
+
+			$icon = $service->getIcon($this->entity, $size);
+
+			$this->assertTrue($icon->exists());
+			$this->assertEquals(filemtime($icon->getFilenameOnFilestore()), $service->getIconLastChange($this->entity, $size));
+			$this->assertEquals(elgg_get_inline_url($icon, true), $service->getIconURL($this->entity, ['size' => $size]));
+
+			$imagesize = getimagesize($icon->getFilenameOnFilestore());
+
+			$square = (bool) $opts['square'];
+			if ($square) {
+				$this->assertTrue($imagesize[0] == $imagesize[1]);
+			}
+
+			$upscale = (bool) $opts['square'];
+			if ($upscale) {
+				$this->assertTrue($imagesize[0] == $opts['w']);
+			} else {
+				$this->assertTrue($imagesize[0] <= $opts['w']);
+			}
+		}
+	}
+
+	/**
+	 * @group IconService
+	 */
+	public function testCanSaveIconFromUploadedFile() {
+
+		$local_file = $this->config->get('dataroot') . '1/1/white.png';
+			
+		$upload = new \Symfony\Component\HttpFoundation\File\UploadedFile($local_file, 'white.png', 'random/foo', filesize($local_file), UPLOAD_ERR_OK, true);
+		$this->request->files->set('icon', $upload);
+
+		$service = $this->createService();
+
+		$sizes = $service->getSizes('object', 'foo');
+		$this->assertNotEmpty($sizes);
+
+		$service->saveIconFromUploadedFile($this->entity, 'icon');
+
+		foreach ($sizes as $size => $opts) {
+
+			$icon = $service->getIcon($this->entity, $size);
+
+			$this->assertTrue($icon->exists());
+			$this->assertEquals(filemtime($icon->getFilenameOnFilestore()), $service->getIconLastChange($this->entity, $size));
+			$this->assertEquals(elgg_get_inline_url($icon, true), $service->getIconURL($this->entity, ['size' => $size]));
+
+			$imagesize = getimagesize($icon->getFilenameOnFilestore());
+
+			$square = (bool) $opts['square'];
+			if ($square) {
+				$this->assertTrue($imagesize[0] == $imagesize[1]);
+			}
+
+			$upscale = (bool) $opts['square'];
+			if ($upscale) {
+				$this->assertTrue($imagesize[0] == $opts['w']);
+			} else {
+				$this->assertTrue($imagesize[0] <= $opts['w']);
+			}
+		}
+	}
+
+	/**
+	 * @group IconService
+	 */
+	public function testSaveIconWithCroppingCoordinates() {
```

Comment:
Need to do research on how others test cropping. I imagine we could generate a black&white image and compare by reading bytes.